### ISSUE TITLE
Show "-1 confirmations" for orphaned Omni transactions

### ISSUE THREAD

**Issue by dexX7** [OmniLayer/omnicore#244]
The behavior of `"getrawtransaction"`:

- confirmed transactions have `n` confirmations
- unconfirmed transactions have `0` confirmations
- orphaned transactions have `0` confirmations

The behavior of `"gettransaction"`:

- confirmed transactions have `n` confirmations
- unconfirmed transactions have `0` confirmations
- orphaned transactions have `-1` confirmations

The behavior of `"omni_gettransaction"`:

- confirmed transactions have `n` confirmations and are `valid`
- unconfirmed transactions have `0` confirmations and are not `valid`
- orphaned transactions have `n` confirmations and are not `valid`

For the sake of consistency, orphaned Omni transactions should not have any confirmations, but instead `0` or `-1` confirmations, to make it more clear that the transaction doesn't refer to the active chain.

**Comment by zathras-crypto** [comment]
Can I clarify?

* Block 5 tx abc123 is mined
* tx abc123 shows 1 confirmation
* Block 5 is invalidated and replaced with a new longer chain of blocks 5 & 6, none of which contain the tx (so it's now back in mempool)
* tx abc123 shows 2 confirmations

Is that accurate?

**Comment by zathras-crypto** [comment]
Note in your behavior above you say that orphaned transactions show -1 in '''gettransaction''' - are you sure?  Transactions in orphaned blocks should go back to the mempool and show 0 confirmations (only conflicted transactions should show -1 which indicates it will never confirm).

**Comment by dexX7** [comment]
Yes, that's correct. We currently don't differentiate between "confirmations" and "confirmations of blocks in the active chain" in [rpctxobject.cpp#L51-L58](https://github.com/OmniLayer/omnicore/blob/d62b0a964924b3804fb5c6805b37e401dd20052c/src/omnicore/rpctxobject.cpp#L51-L58).

> Note in your behavior above you say that orphaned transactions show -1 in '''gettransaction''' - are you sure?

Yes, see [GetDepthInMainChain()](https://github.com/OmniLayer/omnicore/blob/b6347bf8138395117fa6d3a753c8aa266b05ada0/src/wallet.cpp#L2328) is used for `"gettransaction"` (wallet transactions).

> Transactions in orphaned blocks should go back to the mempool and show 0 confirmations

Ah sorry, in this case I was referring to "really invalidated and orphaned" transactions, which are not even in the mempool anymore. Note that these are still in the (Core) TXDB (with a reference to the orphaned block).

**Comment by dexX7** [comment]
As a side note: for the reorg tests I use `invalidateblock` followed by `clearmempool`, to get rid of all leftovers and to ensure there are no unexpected side effects from transactions, which were intended to be removed, but made it back via the mempool.

For example, let's say we want to test that the balances after a reorg are correct, and the invalidated transactions are not "counted", then I'd:

1. send some tokens
2. mine block X
3. remember balance
4. send some other tokens (to be reverted!)
5. mine block Y
6. invalidate block Y
7. clear mempool
8. mine block Z
9. check that balance == initial balance of step 2

In this case step 7 reverts step 4, and without it, the second send would be picked up via block Z.

**Comment by zathras-crypto** [comment]
> Ah sorry, in this case I was referring to "really invalidated and orphaned" transactions, which are not even in the mempool anymore. Note that these are still in the (Core) TXDB (with a reference to the orphaned block).

Ahhh wording again hehe :)  I always thought:
* an orphaned transaction was a perfectly valid transaction that had it's block orphaned and was waiting to be re-mined.  
* an invalid transaction was a transaction that was disallowed for some reason (bad sig, inputs already spent etc etc)

> example

Sorry for the confusion, but if the underlying Bitcoin transaction was valid in step 4, what makes it invalid later on?

**Comment by zathras-crypto** [comment]
> In this case steps 6, 7, 8 revert step 4, and without it, the second send would be picked up via block Z.

I guess this is my confusion - because isn't that what would actually happen?  If the block containing the second send is orphaned, then the second send would be mined in a subsequent block.  If it was a valid Bitcoin transaction before the block was orphaned then surely it's a valid transaction after the block was orphaned (assuming the reorg doesn't invalidate the inputs).

Sorry if I'm just not getting it hehe

**Comment by zathras-crypto** [comment]
I see what you're doing here - reading https://github.com/OmniLayer/OmniJ/commit/290e69b0dacf01f3acbb1d4e0b86580e707aeeab helped me understand :)

**Comment by dexX7** [comment]
> an orphaned transaction ... an invalid transaction ...

Sounds good to me. I guess in your example description it comes down to the question whether an orphaned transaction is valid. I don't think so ("no effect was applied"), but similar to unconfirmed transactions, the question whether the transaction is "valid" may not be the right question at all, and if we only consider structual-validity, then it's a different story.

> I guess this is my confusion - because isn't that what would actually happen? If the block containing the second send is orphaned, then the second send would be mined in a subsequent block.

Due to the mempool clearing, it's as if it never happened. This would be similar to:

> assuming the reorg ~~doesn't~~ invalidate[s] the inputs

Sorry for the confusion, let's tackle it this way, and clarify:

- when should we show `"confirmations"`:
 - for unconfirmed transactions? (yes: `0`)
 - for confirmed transactions? (yes: `n`)
 - for orphaned transactions? (???)

- when we should show `"valid"` (either `true` or `false`):
 - for unconfirmed transactions? (currently not shown)
 - for confirmed transactions? (yes: `true|false`)
 - for orphaned transactions? (???)

- when we should show `block`, `blockhash`, `blocktime`:
 - for unconfirmed transactions? (currently not shown)
 - for confirmed transactions? (yes)
 - for orphaned transactions? (???)

To clarify: when I refer to orphaned transactions, then I mean transactions which are not in the mempool (otherwise I'd consider them as unconfirmed).

**Comment by sukhdev22** [comment]
What happens to an "INVALID" transaction due to "INSUFFICIENT FUND" ??

**Comment by dexX7** [comment]
@sukhdev22: please clarify what you are asking exactly. A transaction, e.g. a send, can fail, if the sender doesn't have enough balance. In this case no tokens are moved.
